src/FontManager.cc

```diff
@@ -1,251 +1,265 @@
-#include <stdlib.h>
-#include <node.h>
-#include <uv.h>
-#include <v8.h>
-#include <nan.h>
-#include "FontDescriptor.h"
+#include <cstdlib>
+
+#include "napi.h"
 
-using namespace v8;
+#include "FontDescriptor.h"
+#include "ResultSet.h"
+#include "FontManagerImpl.h"
 
-// these functions are implemented by the platform
-long getAvailableFonts(ResultSet **);
-long findFonts(ResultSet **, FontDescriptor *);
-long findFont(FontDescriptor **, FontDescriptor *);
-long substituteFont(FontDescriptor **, char *, char *);
+using namespace Napi;
 
 // converts a ResultSet to a JavaScript array
-Local<Array> collectResults(ResultSet *results) {
-  Nan::EscapableHandleScope scope;
-  Local<Array> res = Nan::New<Array>(results->size());
+Napi::Array collectResults(Napi::Env env, ResultSet *results) {
+  Napi::EscapableHandleScope scope(env);
+  Napi::Array res = Napi::Array::New(env, results->size());
 
   int i = 0;
   for (ResultSet::iterator it = results->begin(); it != results->end(); it++) {
-    Nan::Set(res, i++, (*it)->toJSObject());
+    res.Set(i++, (*it)->toJSObject(env));
   }
 
   delete results;
-  return scope.Escape(res);
+  return scope.Escape(res).As<Napi::Array>();
 }
 
 // converts a FontDescriptor to a JavaScript object
-Local<Value> wrapResult(FontDescriptor *result) {
-  Nan::EscapableHandleScope scope;
+Napi::Value wrapResult(Napi::Env env, FontDescriptor *result) {
+  Napi::EscapableHandleScope scope(env);
   if (result == NULL)
-    return scope.Escape(Nan::Null());
+    return scope.Escape(env.Null());
 
-  Local<Object> res = result->toJSObject();
+  Napi::Object res = result->toJSObject(env);
   delete result;
   return scope.Escape(res);
 }
 
-// holds data about an operation that will be
-// performed on a background thread
-struct AsyncRequest {
-  uv_work_t work;
-  FontDescriptor *desc;     // used by findFont and findFonts
-  char *postscriptName;     // used by substituteFont
-  char *substitutionString; // ditto
-  FontDescriptor *result;   // for functions with a single result
-  ResultSet *results;       // for functions with multiple results
-  long error;               // result/results is defined if error == 0
-  Nan::Callback *callback;  // the actual JS callback to call when we are done
-
-  AsyncRequest(Local<Value> v) {
-    work.data = (void *)this;
-    callback = new Nan::Callback(v.As<Function>());
-    desc = NULL;
-    postscriptName = NULL;
-    substitutionString = NULL;
-    result = NULL;
-    results = NULL;
-  }
-
-  ~AsyncRequest() {
-    delete callback;
-
-    if (desc)
-      delete desc;
+// handles asynchrnous execution for interface functions
+// this destroys itself unless we call AsyncWorker::SuppressDestruct()
+template<typename _ReturnType>
+struct PromiseAsyncWorker : public Napi::AsyncWorker {
+  static_assert(std::is_same<_ReturnType, ResultSet>::value || std::is_same<_ReturnType, FontDescriptor>::value, "");
```

Comment:
why not add an error message instead of ""?

Reply:
in my opinion, static asserts like this should usually be self-documenting. and the empty message was because message-less static asserts aren't supported in whatever version of c++ this is(14 i think?).

I can add a message if you guys don't agree though.

Reply:
I think you are right about being self-documenting